include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,189 @@
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams, currInteraction, n, ny, a2, ay2);
+			float VdotH = microfacet.isotropic.VdotH;
+			float VdotH_clamp = irr_glsl_conditionalAbsOrMax(is_bsdf, VdotH, 0.0);
+			vec3 fr;
+#ifdef OP_CONDUCTOR
+			if (op == OP_CONDUCTOR)
+				fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH_clamp);
+			else
+#endif
+				fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH_clamp);
+
+			const float NdotL = cosFactor;
+			const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
+			if (is_bsdf)
+			{
+				float eta = ior_scalar;
+				float orientedEta, rcpOrientedEta;
+				irr_glsl_getOrientedEtas(orientedEta, rcpOrientedEta, currInteraction.isotropic.NdotV, eta);
+
+				float LdotH = microfacet.isotropic.LdotH;
+				float VdotHLdotH = VdotH * LdotH;
+#ifdef NDF_GGX
+				if (ndf == NDF_GGX)
+					bxdf_eval_scalar_part = irr_glsl_ggx_microfacet_to_light_measure_transform(bxdf_eval_scalar_part, NdotL, refraction, VdotH, LdotH, VdotHLdotH, orientedEta);
+				else
 #endif
+					bxdf_eval_scalar_part = irr_glsl_microfacet_to_light_measure_transform(bxdf_eval_scalar_part, NdotV, refraction, VdotH, LdotH, VdotHLdotH, orientedEta);
+			}
 
-
-#ifndef ONLY_ONE_NDF
+			bxdf_eval = (refraction ? (vec3(1.0)-fr):fr) * bxdf_eval_scalar_part;
 		} else
 #endif
-#endif
-
-#ifndef ONLY_ONE_NDF
-		{} //else "empty braces"
-#endif
+		{}
 	}
 
-	uvec3 regs = instr_decodeRegisters(instr);
-#ifdef OP_DIFFUSE
-	if (op==OP_DIFFUSE) {
-		instr_execute_cos_eval_DIFFUSE(instr, regs, params, bsdf_data);
-	} else
-#endif
-#ifdef OP_CONDUCTOR
-	if (op==OP_CONDUCTOR) {
-		instr_execute_cos_eval_CONDUCTOR(instr, regs, bxdf_eval_scalar_part, params, bsdf_data);
-	} else
-#endif
-#ifdef OP_PLASTIC
-	if (op==OP_PLASTIC) {
-		instr_execute_cos_eval_PLASTIC(instr, regs, bxdf_eval_scalar_part, params, bsdf_data);
-	} else
-#endif
+	bxdf_eval_t result = bxdf_eval;
+	if (!op_isBXDF(op))
+	{
+		mat2x4 srcs = instr_fetchSrcRegs(instr, regs);
+
+		BEGIN_CASES(op)
 #ifdef OP_COATING
-	if (op==OP_COATING) {
-		instr_execute_cos_eval_COATING(instr, regs, params, bsdf_data);
-	} else
-#endif
-#ifdef OP_DIFFTRANS
-	if (op==OP_DIFFTRANS) {
-		instr_execute_cos_eval_DIFFTRANS(instr, regs, params, bsdf_data);
-	} else
-#endif
-#ifdef OP_DIELECTRIC
-	if (op==OP_DIELECTRIC) {
-		instr_execute_cos_eval_DIELECTRIC(instr, regs, params, bsdf_data);
-	} else
-#endif
-#ifdef OP_THINDIELECTRIC
-	if (op==OP_THINDIELECTRIC) {
-		instr_execute_cos_eval_THINDIELECTRIC(instr, regs, params, bsdf_data);
-	} else
+		CASE_BEGIN(op, OP_COATING) {
+			float dummy;
+			result = instr_execute_cos_eval_COATING(instr, srcs, params, ior[0], ior2[0], s, microfacet, bsdf_data, dummy);
+		} CASE_END
 #endif
 #ifdef OP_BLEND
-	if (op==OP_BLEND) {
-		instr_execute_cos_eval_BLEND(instr, regs, params, bsdf_data);
-	} else
+		CASE_BEGIN(op, OP_BLEND) {
+			result = instr_execute_cos_eval_BLEND(instr, srcs, params, bsdf_data);
```

Comment:
so coating is still not unified with blend?

Reply:
it is blend-like, but has different opcode. Honestly i cant see any point in unifying them further, in the end of the day they differ a lot (first of all blend takes weight from bsdf_data or texture and coating computes it). So there would be a need for flag determining how weights are obtained which already stands for divergence between most of the work done by the opcode

Reply:
reading and writing data to random registers might be expensive... so it would be nice not to diverge for that

Reply:
hmm ok I'll allow a COAT which is a different opcode to BLEND

Reply:
ok I guess its fine